### ISSUE TITLE
"Equals := 0 = MEMCMP(THIS, ADR(Station), SIZEOF(FB_Station));" breaks the formatter

### ISSUE THREAD

**Issue by GISED-Link** [Barteling/STweep#38]
**Describe the bug**
Do not format anything

 - The initial code before formatting.
```
Equals := FALSE;
Equals           := FALSE; // this is stupide, it is just to be sure that STweep runs correctly or not




IF __ISVALIDREF(Station) THEN
	Equals := 0 = MEMCMP(THIS, ADR(Station), SIZEOF(FB_Station));
		END_IF
```
 - The current result after formatting.
 nothing changed

 - The expected result.
```
Equals := FALSE;
Equals := FALSE; // this is stupide, it is just to be sure that STweep runs correctly or not

IF __ISVALIDREF(Station) THEN
    Equals := 0 = MEMCMP(THIS, ADR(Station), SIZEOF(FB_Station));
END_IF
```
 - The folowing code does format correctly (commented the line with the MEMCMP):
```
Equals := FALSE;

IF __ISVALIDREF(Station) THEN
    //Equals := 0 = MEMCMP(THIS, ADR(Station), SIZEOF(FB_Station));
    Equals := TRUE;
END_IF
```

 - An export of the used settings.
[GISED_Link_STWEEP_config.zip](https://github.com/Barteling/STweep/files/6278147/GISED_Link_STWEEP_config.zip)

**STweep version info**
STweep for TwinCAT 3.3.0-Beta

**Comment by GISED-Link** [comment]
EDITED
To compare 2 FB of the same type, I tried to do this:
```
IF ADR(THIS^._MoverList[i].Station) = ADR(StationTest) THEN
    // do some code here
END_IF
```
where `THIS^._MoverList[i].Station` is a member of the FB (I know it is obvious) and is a reference to a FB_Station. 
where `StationTest` is an input of my POU and a reference to FB_Station. 

The compiler gives me this error:
`THIS^._MoverList[i].Station' is not allowed as operand for ADR`
but nothing for StationTest.

I have not idea if it is totally normal or not, but I think this is strange.
By creating a method "Equals" in my FB_Station, I am now able to compare the 2 FB togheter (and it is really better).

The question is: Is it enough to compare the 2 adresses or do I need to do a proper "memcpy"? Or what is the best to do? 
In my case, only comparing the adresses is enough and (should be) better. But what is the common way to do this?

**Comment by Barteling** [comment]
Thanks for reporting I'll check what problem is.


So regarding your FB comparison, what are you trying to find out, that you have two fb's which are the same instance? 
Or that you have two FB's which are of the same type? perhaps derived. 

Is the __VarInfo perhaps of use?

https://infosys.beckhoff.com/english.php?content=../content/1033/tc3_plc_intro/2529165707.html&id=

**Comment by Barteling** [comment]
<pre>
	Equals := 0 = MEMCMP(THIS, ADR(Station), SIZEOF(FB_Station));
</pre>

The problem is the 'THIS' argument. STweep expects the dereference symbol '^' after 'THIS'.

Fails to format<pre>MethodCall(THIS);</pre> 
Formats ok<pre>MethodCall(THIS^);</pre> 

I'll update the parser.

**Comment by GISED-Link** [comment]
> Thanks for reporting I'll check what problem is.
> 
> So regarding your FB comparison, what are you trying to find out, that you have two fb's which are the same instance?
> Or that you have two FB's which are of the same type? perhaps derived.
> 
> Is the __VarInfo perhaps of use?
> 
> https://infosys.beckhoff.com/english.php?content=../content/1033/tc3_plc_intro/2529165707.html&id=

Thank you for the input. I never used __VarInfo berfore but it looks nice.

What I'm trying to do is to replicate the c# object.Equals method. I think I have found something in the c# doc:
https://docs.microsoft.com/en-us/dotnet/api/system.object.equals?redirectedfrom=MSDN&view=net-5.0#System_Object_Equals_System_Object_System_Object_
> **Remarks**
> The static Equals(Object, Object) method indicates whether two objects, objA and objB, are equal. It also enables you to test objects whose value is null for equality. It compares objA and objB for equality as follows:
> 
> - It determines whether the two objects represent the same object reference. If they do, the method returns true. This test is equivalent to calling the ReferenceEquals method. In addition, if both objA and objB are null, the method returns true.
> - It determines whether either objA or objB is null. If so, it returns false.
> - If the two objects do not represent the same object reference and neither is null, it calls objA.Equals(objB) and returns the result. This means that if objA overrides the Object.Equals(Object) method, this override is called.

For me, checking the references is enough. But it looks like the c# implementation goes a step further by checking the (maybe overloaded) method Equals. Since Twincat allows the overloading, it would be a nice idea to force all my FB to have a default Equals method and overload it when necessary...
https://infosys.beckhoff.com/english.php?content=../content/1033/tc3_plc_intro/3537333899.html&id=
I would love to see a post about this topic on your PLCopen blog 😃

**Comment by Roald87** [comment]
@GISED-Link You could use the [AssertEquals](https://tcunit.org/api/#AssertEquals) method from the TcUnit library.

**Comment by GISED-Link** [comment]
@Roald87 Thanks of the idea, I have not thought about it.

I have looked at the function METHOD PUBLIC AssertEquals found here:
https://github.com/tcunit/TcUnit/blob/master/TcUnit/TcUnit/POUs/FB_TestSuite.TcPOU

I found a MASSIVE implementation 😮, a bit too much for what I intended. The "test_equal" schem is really good, it really inspire me 🥇

By the way, the only twincat forum I found is on linkedin... and I'm not accepted yet. Is it possible that we create one? There are (at least) 3 really good and active blogs (https://www.plccoder.com/, https://stefanhenneken.net/, https://alltwincat.com/) but none of them have a "forum" section. Even Siemens have it... Since Twincat uses the Codesys compiler, is it possible to create a forum where both plateform are covered? Obviously I can create it, but it will make no sense if I'm the only one ... I could also create a another blog with a forum ... but I would prefer to connect blogs together on the same forum plateform...

**Comment by Roald87** [comment]
> I have looked at the function METHOD PUBLIC AssertEquals found here:

That's the same one.

> I found a MASSIVE implementation 😮, a bit too much for what I intended. The "test_equal" schem is really good, it really inspire me 

You can just use the TcUnit library as well. No need to write your own implementation.

> By the way, the only twincat forum I found is on linkedin... and I'm not accepted yet. 

Just wait, you'll probably be accepted. I think it is the best place for TwinCAT related discussions. No need to start another one in my opinion. Although you can always do it of course :p. Also more and more stuff gets posted on StackOverflow, so you might be able to find answers/ask quetions there as well.

**Comment by GISED-Link** [comment]
> You can just use the TcUnit library as well. No need to write your own implementation.

This code throw a message, I just want to return a BOOL value and if possible to have a base abstract class that all my FB derive from so they get an already implemented Equals method (and a ToString()). Or I am missunderstanding something?

**Comment by Aliazzzz** [comment]
EDIT: fixed wikipedia url

A forum is a good idea, but for sake of completeness I wish to inform you that TwinCAT users are more then welcome on to post questions on www.forge.codesys.com. The codesys forge is open to any CODESYS variant on the market, so TwinCAT is heartly welcome. Forge is a multi faceted forum, with wiki's, git and / or svn repo's for open source projects, faq, support etc, etc. It is built using Apache Allura Forge platform, so maybe that is an idea to use for a TwinCAT oriented all in one forum and open source git project hosting platform? To read up on on the forge concept, it offers so much more then a simple forum! [https://en.wikipedia.org/wiki/Forge_(software)](https://en.wikipedia.org/wiki/Forge_(software))

**Comment by Barteling** [comment]
Fixed in 3.4.0 both TwinCAT and Codesys versie.
